Add chart loading overlay

janklimo opened the pull request:
Hi @forsbergplustwo 👋 Kudos for open-sourcing the app!

I was poking around the codebase to get more familiar with it and came up with this UX improvement:

https://github.com/forsbergplustwo/partner-metrics/assets/7811733/761a6951-e74c-445f-aa66-496bdbb28500

Some actions can take a while so I think some kind of loading indicator would make things more intuitive. 

Let me know what you think!

forsbergplustwo commented on app/javascript/controllers/loading_controller.js:
```diff
@@ -0,0 +1,9 @@
+
+export default class extends Controller {
+  show() {
+    const loadingOverlays = document.querySelectorAll(".loading-overlay");
```
Suggest we use a stimulus target for this. Makes it explicit both in stimulus and html markup what is hooked up to what (one of my favourite things about Stimulus 😊).

forsbergplustwo commented on app/views/metrics/_tile_with_chart.html.erb:
```diff
@@ -1,4 +1,7 @@
-<%= polaris_card(title: t("metrics.tiles.#{tile.handle}", default: tile.handle.to_s.humanize)) do |card| %>
+<%= polaris_card(title: t("metrics.tiles.#{tile.handle}", default: tile.handle.to_s.humanize), classes: 'relative') do |card| %>
+  <div class="loading-overlay hidden Polaris-LegacyCard">
```
Is it possible to avoid hard coding a Polaris css class here? It would be great to either:

a) Use a `polaris_card` component instead
or
b) Extract needed CSS from `.Polaris-LegacyCard` into a custom class.

forsbergplustwo commented on app/views/metrics/_tile_with_chart.html.erb:
```diff
@@ -1,4 +1,7 @@
-<%= polaris_card(title: t("metrics.tiles.#{tile.handle}", default: tile.handle.to_s.humanize)) do |card| %>
+<%= polaris_card(title: t("metrics.tiles.#{tile.handle}", default: tile.handle.to_s.humanize), classes: 'relative') do |card| %>
+  <div class="loading-overlay hidden Polaris-LegacyCard">
+    <%= polaris_spinner(accessibility_label: "Loading chart data...") %>
```
Great to see accessibility labels used. Can you please add `i18n` for the string? Most of the app uses i18n for user facing strings, so would be great to have it here too.

forsbergplustwo reviewed (commented):
Thanks so much for this @janklimo , it's a very welcome addition and helps a lot! 😎

Added a few minor comments, and was thinking overall for the stimulus controller, it currently attaches 3 times and then targets an out of scope element using CSS. Perhaps the stimulus controller can be attached higher in the DOM. For example:

```ruby
# views/metrics/show.html.erb

  <%= turbo_frame_tag, :metrics, 
    data: {
      controller: "loading",
      action: "submit@window->loading#show"
    }
   do %>
```

That brings the loading html element into scope of the stimulus controller so you can add a target. It also removes the need to attach the stimulus controller and action to each form element.

What do you think?

janklimo commented:
Great feedback @forsbergplustwo let me take this back for a bit.

janklimo commented:
@forsbergplustwo I've made these changes:

1. The Stimulus controller works with explicit targets.
2. It is attached to the `#metrics` element. Removed it from everywhere else.
3. It listens to Turbo's `turbo:before-fetch-request` events. I tried making it work by intercepting a submit event but the issue is that the tile actions aren't forms but plain links. I tried to convert them to forms but it didn't work well and ultimately it was ugly code for no good reason. This way, we don't have to listen to click or form submit events. Instead, we intercept a Turbo event before a request is dispatched.
4. Used plain card component for the overlay, added i18n string. This made the custom CSS a bit shorter.

forsbergplustwo reviewed (approved):
Fantastic, great work @janklimo 👏 And you are totally right, the turbo event is a better fit!
